### DISCUSSION TITLE
[@types/dom-chromium-ai] AIAssistantCreateOptionsWithSystemPrompt should not permit the field initialPromps

### DISCUSSION THREAD

**Discussion by wginsberg** [DefinitelyTyped/DefinitelyTyped#70869] [Issues with a @types package]
There is a runtime error thrown when passing both `systemPrompt` and `initialPrompt` to window.ai.assistant.create:
```
window.ai.assistant.create({
    systemPrompt: "You are a dog",
    initialPrompts: [{ role: "user", content: "meow" }]
})
> Uncaught (in promise) TypeError: The systemPrompt and initialPrompts should not present at the same time.
    at <anonymous>:2:37
```

To surface this in the typechecker the interface could be changed from 
```
// current interface
interface AIAssistantCreateOptionsWithSystemPrompt extends AIAssistantCreateOptions {
    systemPrompt?: string;
    initialPrompts?: Array<AIAssistantAssistantPrompt | AIAssistantUserPrompt>;
}
```
to
```
// proposed interface
interface AIAssistantCreateOptionsWithSystemPrompt extends AIAssistantCreateOptions {
    systemPrompt?: string;
    initialPrompts?: never;
}
```

[a comment by an automated account is not shown]

**Comment by christianliebel** [comment]
@wginsberg Thank you for your feedback. The situation is a bit tricky, as specifying both a system prompt and initial prompts is actually an intended use of this API, but its current (experimental) implementation in Chrome may indeed throw a runtime error. Given that this scenario should be supported in the future, I would prefer to keep the typing as-is.

**Comment by tomayac** [comment]
The current limitations are [mentioned in the explainer](https://github.com/explainers-by-googlers/prompt-api/#:~:text=Using%20both%20systemPrompt%20and%20a%20%7B%20role%3A%20%22system%22%20%7D%20prompt%20in%20initialPrompts%2C%20or%20using%20multiple%20%7B%20role%3A%20%22system%22%20%7D%20prompts%2C%20or%20placing%20the%20%7B%20role%3A%20%22system%22%20%7D%20prompt%20anywhere%20besides%20at%20the%200th%20position%20in%20initialPrompts%2C%20will%20reject%20with%20a%20TypeError.):

> Using both systemPrompt and a { role: "system" } prompt in initialPrompts, or using multiple { role: "system" } prompts, or placing the { role: "system" } prompt anywhere besides at the 0th position in initialPrompts, will reject with a TypeError.

@domenic, are those temporary restrictions of the implementation, or do we expect these limitations to remain and therefore worthwhile to catch in the types?  

(Note: The example from the original post is just using `systemPrompt` and `initialPrompts` together, but doesn't otherwise harm any of the limitations from above.)

**Reply by domenic** [reply]
Those are intended to be permanent and eventually written into the spec. (The net effect is only <= 1 system prompt, and always in the 0th position.) I think they match other APIs that use system prompts, but if I'm missing something, an issue on the explainer repo would be good.

**Reply by domenic** [reply]
But note that the error in the OP does not match the explainer's limitations. Using systemPrompt and a non-`{ role: "system" }` initialPrompt should be fine. That's worth filing a Chromium bug for.

**Reply by tomayac** [reply]
> Those are intended to be permanent and eventually written into the spec. (The net effect is only <= 1 system prompt, and always in the 0th position.) I think they match other APIs that use system prompts, but if I'm missing something, an issue on the explainer repo would be good.

Thank you, @domenic! @wginsberg and @christianliebel, I think it makes sense then to bake those limitations into the types.
  
> But note that the error in the OP does not match the explainer's limitations. Using systemPrompt and a non-`{ role: "system" }` initialPrompt should be fine. That's worth filing a Chromium bug for.

Yes, also noted this in my comment. @wginsberg, could I ask you to submit a quick [new Chromium bug](https://new.crbug.com) with this error case? Thanks!

**Reply by christianliebel** [reply]
@tomayac I think this restriction is already part of the current types:

```ts
interface AIAssistantCreateOptionsWithSystemPrompt extends AIAssistantCreateOptions {
    systemPrompt?: string;
    initialPrompts?: Array<AIAssistantAssistantPrompt | AIAssistantUserPrompt>;
}

interface AIAssistantCreateOptionsWithoutSystemPrompt extends AIAssistantCreateOptions {
    systemPrompt?: never;
    initialPrompts?:
        | [AIAssistantSystemPrompt, ...Array<AIAssistantAssistantPrompt | AIAssistantUserPrompt>]
        | Array<AIAssistantAssistantPrompt | AIAssistantUserPrompt>;
}
```
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/62a8d0c3371bd5c2b9b482be4f9b496fed514416/types/dom-chromium-ai/index.d.ts

**Reply by tomayac** [reply]
OK then! Anything else the types don't cover? Or are we good? 

(Note that this proposal, if still relevant, needs to reflect `self.ai.assistant` → `self.ai.languageModel`.)

**Reply by christianliebel** [reply]
I think we're good!

**Reply by wginsberg** [reply]
> > Those are intended to be permanent and eventually written into the spec. (The net effect is only <= 1 system prompt, and always in the 0th position.) I think they match other APIs that use system prompts, but if I'm missing something, an issue on the explainer repo would be good.
> 
> Thank you, @domenic! @wginsberg and @christianliebel, I think it makes sense then to bake those limitations into the types.
> 
> > But note that the error in the OP does not match the explainer's limitations. Using systemPrompt and a non-`{ role: "system" }` initialPrompt should be fine. That's worth filing a Chromium bug for.
> 
> Yes, also noted this in my comment. @wginsberg, could I ask you to submit a quick [new Chromium bug](https://new.crbug.com) with this error case? Thanks!

Filed a chromium bug

https://issues.chromium.org/issues/373642641
